File: backend/api/v1/notifications.py
"""
Notifications API

Endpoints for in-app notification management:
- GET  /notifications              — list unread notifications
- GET  /notifications/count        — unread count (for bell badge)
- PUT  /notifications/read-all     — mark all notifications read
- PUT  /notifications/{id}/read    — mark a single notification read
"""

import uuid

from fastapi import APIRouter, Depends, HTTPException, Path
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import get_current_user, get_db_session, SessionData
from services.notification_service import NotificationService

router = APIRouter(prefix="/notifications", tags=["Notifications"])


@router.get("")
async def get_notifications(
    current_user: SessionData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_session),
):
    """Return the caller's unread notifications (newest first, max 50)."""
    svc = NotificationService(db)
    items = await svc.get_unread(current_user.user_id)
    return {"notifications": items, "total": len(items)}


@router.get("/count")
async def get_notification_count(
    current_user: SessionData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_session),
):
    """Return the number of unread notifications for the caller."""
    svc = NotificationService(db)
    count = await svc.get_unread_count(current_user.user_id)
    return {"unread": count}


@router.put("/read-all")
async def mark_all_notifications_read(
    current_user: SessionData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_session),
):
    """Mark all unread notifications as read for the caller."""
    svc = NotificationService(db)
    count = await svc.mark_all_read(current_user.user_id)
    return {"success": True, "marked": count}


@router.put("/{notification_id}/read")
async def mark_notification_read(
    notification_id: uuid.UUID = Path(..., description="UUID of the notification to mark read"),
    current_user: SessionData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_session),
):
    """Mark a single notification as read. Returns 404 if not found or already read."""
    svc = NotificationService(db)
    success = await svc.mark_read(current_user.user_id, str(notification_id))
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"success": True}

File: frontend/lib/api/notifications.ts
/**
 * Notifications API functions
 *
 * CRUD operations for in-app user notifications:
 * - list unread notifications
 * - get unread count (for bell badge polling)
 * - mark individual notification as read
 * - mark all notifications as read (client-side batch)
 */

import { apiClient } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Notification {
  id: string
  type: string
  title: string
  body: string | null
  created_at: string
}

export interface GetNotificationsResponse {
  notifications: Notification[]
  total: number
}

export interface GetNotificationCountResponse {
  unread: number
}

export interface MarkReadResponse {
  success: boolean
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * Get unread notifications for the current user (newest first, max 50)
 */
export async function getNotifications(): Promise<GetNotificationsResponse> {
  return apiClient.get<GetNotificationsResponse>('/notifications')
}

/**
 * Get the unread notification count for the current user.
 * Designed to be polled frequently (e.g. every 30s) for the bell badge.
 */
export async function getNotificationCount(): Promise<GetNotificationCountResponse> {
  return apiClient.get<GetNotificationCountResponse>('/notifications/count')
}

/**
 * Mark a single notification as read.
 * Returns 404 if the notification does not exist or is already read.
 */
export async function markNotificationRead(id: string): Promise<MarkReadResponse> {
  return apiClient.put<MarkReadResponse>(`/notifications/${id}/read`, {})
}

/**
 * Mark all currently unread notifications as read.
 *
 * The backend does not expose a bulk-mark-all endpoint, so this function
 * fetches the current list and marks each one individually.  It settles all
 * requests in parallel and ignores individual failures (e.g. 404 for a
 * notification that was already read by another tab).
 */
export async function markAllRead(): Promise<void> {
  const data = await getNotifications()
  await Promise.allSettled(
    data.notifications.map((n) => markNotificationRead(n.id))
  )
}
